openreview/matching.py

```diff
@@ -80,3 +82,294 @@ def get_domains(entity, subdomains = False):
+
+    score_matrix, hard_constraint_dict, user_by_index, forum_by_index = encode_score_matrix(scores_by_user_by_forum)
+
+    solution = Solver(alphas, betas, score_matrix, hard_constraint_dict).solve()
+
+    assigned_userids = decode_score_matrix(solution, user_by_index, forum_by_index)
+
+    new_assignment_notes = create_or_update_assignments(
+        assigned_userids,
+        existing_assignments,
+        assignment_invitation,
+        configuration_note,
+        scores_by_user_by_forum
+    )
+
+    return new_assignment_notes
+
+def get_weighted_scores(metadata_notes, weights, constraints, match_group):
+    '''
+    Returns a dict of dicts that contains weighted feature scores per user, per forum.
+
+    e.g. {
+        'abcXYZ': {
+            '~Michael_Spector1': {
+                'affinityScore': 0.85
+            },
+            '~Melisa_Bok1': {
+                'affinityScore': 0.93
+            }
+        }
+    }
+
+    '''
+
+    scores_by_user_by_forum = {}
+    for m in metadata_notes:
+        user_entries = [e for e in m.content['groups'][match_group.id] if e['userId'] in match_group.members]
+        features_by_user = { entry['userId']: weight_scores(entry['scores'], weights) for entry in user_entries}
+
+        # apply user-defined constraints
+        if m.forum in constraints:
+            constraint_by_user = { user: {'userConstraint': value} for user, value in constraints[m.forum].iteritems()}
+            features_by_user.update(constraint_by_user)
+
+        scores_by_user_by_forum[m.forum] = features_by_user
+
+    return scores_by_user_by_forum
+
+def weight_scores(scores, weights):
+    '''
+    multiplies feature values by weights, excepting hard constraints
+    '''
+    weighted_scores = {}
+    for feature in weights:
+        if feature in scores:
+            weighted_scores[feature] = scores[feature]
+
+            if scores[feature] != '-inf' and scores[feature] != '+inf':
+                weighted_scores[feature] *= weights[feature]
+
+    return weighted_scores
+
+def create_assignment(forum, label, assignment_invitation):
+    '''
+    Creates the JSON record for an empty assignment Note.
+
+    *important* return type is dict, not openreview.Note
+    '''
+
+    return {
+        'forum': forum,
+        'invitation': assignment_invitation.id,
+        'readers': assignment_invitation.reply['readers']['values'],
+        'writers': assignment_invitation.reply['writers']['values'],
+        'signatures': assignment_invitation.reply['signatures']['values'],
+        'content': {
+            'label': label
+        }
+    }
+
+def create_or_update_config(client, label, params, number=None):
+    '''
+    makes a network call to get any existing configuration notes that match the given label,
+    and updates the note with the given params.
+
+    Returns an openreview.Note object with the updated params.
+    '''
+    config_notes = tools.get_all_notes(client, params['invitation'])
```

Comment:
you can get all the notes by label:

/notes?content.label=label

and if number is present you can add number to the request:

/notes?content.label=label&number=number